security/fma-l1-upgrades.md

```diff
@@ -0,0 +1,274 @@
+detect. Strong emphasis should be placed on mitigation.
+
+#### **Recovery Path(s)**:
+
+It may be impossible to recover from this, depending on which storage changes have been made.
+
+Perhaps now is the time to consider adding the Recover Module.
+
+#### Action items:
+
+- [ ] Implement static analysis based checks to ensure that no `SSTORE` operations occur
+      within `OPCM.upgrade()`.
+
+### FM4: Failure to follow upgrade path
+
+#### **Description:**
+
+For an OP Chain upgrading though multiple releases, there is a risk that the correct upgrade path is not followed.
+
+For example, skipping from v2.0.0 to v2.2.0 without going through v2.1.0.
+
+```mermaid
+graph LR
+    A[2.0.0] --> B[2.2.0]
+
+    %% Invisible elements to help center 2.1.0
+    A ~~~C[2.1.0]
+    C ~~~B
+```
+
+This is a particular risk we have chosen to accept as opposed to enforcing the upgrade path onchain,
+because doing so comes with operational [complications](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/op-contracts-manager-single-release-redesign.md#managing-patches-for-opcm-deployments).
+
+#### **Risk Assessment:**
+
+Medium impact, Medium likelihood
+
+#### **Mitigations:**
+
+1. Define the correct upgrade path in the `superchain-registry`.
+2. Enforce the correct upgrade path in `op-deployer` which is the official interface for upgrading OP Chains.
+
+#### **Detection:**
+
+**Monitoring:** We could consider introducing monitoring which:
+
+1. Reads OPCM addresses from the `superchain-registry`
+2. Monitors those addresses for `Upgraded` events
+3. Detects when the version is not correctly incremented.
+
+However, such monitoring would likely be difficult to maintain, would only focus on the 'official' superchain, and is unlikely to be triggered as effective mitigations are not difficult to implement.
+
+#### **Recovery Path(s):**
+
+The recovery path is highly dependent on the nature of the failure resulting from an invalid upgrade path, which is very difficult to predict.
+
+#### Action items:
+
+- [ ] Define the correct upgrade path in the `superchain-registry`.
+- [ ] Enforce the correct upgrade path in `op-deployer` which is the official interface for upgrading OP Chains.
+
+### FM5: Gas Limit Exceeded
+
+#### **Description:**
+
+As the number of OP Chains grows, upgrades might exceed block gas limits, causing failed upgrades.
+
+#### **Risk Assessment:**
+
+Low impact, Medium likelihood
+
+#### **Mitigations:**
+
+Gas benchmarks should be created to determine the full cost of upgradings all OP Chains. This will require
+a dependency on the `superchain-registry` to track the number of OP Chains.
+
+#### **Detection:**
+
+This will be detected when an upgrade fails due to OOG. That's fine.
+
+#### **Recovery Path(s):**
+
+The upgrade will need to be gas optimized or rearchitected to break it up. We can cross that bridge when we get to it.
+
+#### Action items:
+
+- [ ] Create gas benchmarks for upgrading all OP Chains.
```

Comment:
Curious how you'd go about doing this? Have you a simple example to help me understand?

Reply:
We have `test_upgrade_succeeds` in the Monorepo, which currently has a cost of about 360k gas. We could record the gas costs of running it and revert if it gets beyond some value.

So, we know its pretty far off, but we can create an issue for future work to automatically simulate an upgrade of standard chains.

Reply:
Captured by this test: https://github.com/ethereum-optimism/optimism/issues/13835